src/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets/SocketConnection.cs

```diff
@@ -106,17 +109,45 @@ internal SocketConnection(Socket socket, SocketTransport transport)
                     if (result.IsCompleted)
                     {
                         // Pipe consumer is shut down, do we stop writing
-                        _socket.Shutdown(SocketShutdown.Receive);
                         break;
                     }
                 }
 
+                _connectionContext.Abort(ex: null);
                 _input.Complete();
             }
             catch (Exception ex)
             {
-                _connectionContext.Abort(ex);
-                _input.Complete(ex);
+                Exception error = null;
+
+                if (ex is SocketException se)
+                {
+                    if (se.SocketErrorCode == SocketError.ConnectionReset)
+                    {
+                        // Connection reset
+                        error = new ConnectionResetException(ex.Message, ex);
+                    }
+                    else if (se.SocketErrorCode == SocketError.OperationAborted)
+                    {
+                        error = new TaskCanceledException("The request was aborted");
+                    }
+                }
+
+                if (ex is ObjectDisposedException)
```

Comment:
Nit: if this is changed to an `else if`, the last `else if (error == null)` could be changed to just an `else`.